### ISSUE TITLE
Validation - use of wildcarded parameter in successes()/fails() causes compile problems (corresponding Either code worked fine)

### ISSUE THREAD

**Issue by Alex-Ikanow** [functionaljava/functionaljava#152]
I don't _think_ I'm being too stupid here?

I was converting a bunch of code that used Either to use Validation instead

So I was moving:

    final fj.data.List<Either<Throwable, String>> l = fj.data.List.iterableList(ret_val); // (ret_val is java.util.List<Either<Throwable, String>>)
    final fj.data.List<Throwable> errs = Either.lefts(l);

to

    final fj.data.List<Validation<Throwable, String>> l = fj.data.List.iterableList(ret_val); // (ret_val is java.util.List<Validation<Throwable, String>>)
    final fj.data.List<Throwable> errs = Validation.fails(l);

But it all went terribly wrong because the signature of lefts() is

    <A, E> static List<E> lefts(List<Either<E, A>> l); // this works

whereas the signature of fails() is

    <A, E> static List<E> fails(List<Validation<E, ?>> l); // this does not work without lots of work, see below

The problem seems to be that nested wildcards cause lots of problems (see: http://www.angelikalanger.com/GenericsFAQ/FAQSections/TypeArguments.html#FAQ104), eg you cannot do

    fj.data.List<Validation<Throwable, ?>> l2 = l; // Type mismatch

you have to do

    fj.data.List<? extends Validation<Throwable, ?>> l2 = l;

    //Validation.fails((ll2); //  still doesn't work, but ...

    Validation.fails((fj.data.List<Validation<Throwable, ?>>) ll2); // works! (with a warning)

Is there a reason why you can't use the signature from lefts/rights? That works immediately with no type conversion and no warnings?

**Comment by jbgi** [comment]
Yep looks like Validation need some love. We cannot just apply the fix though, because it breaks other parts of Validation (accumulate methods), which would need to be fixed.
Also the FailProjection inner class need to be static, and accumulate methods should have used NonEmptyList instead of List for errors.

**Comment by Alex-Ikanow** [comment]
Oh I see why you used a ? now - it enables you to accumulate errors without caring about the value type - that's nice, too bad Java's type inference isn't up to the job! What about if you just make the fails()  signature be:

    <A, E> static List<E> fails(List<? extends Validation<E, ?>> l);

and then stick a comment in the Javadoc explaining that a conversion of the above type is needed. I don't think that can break any existing code and it will make it much less mysterious.

**Comment by mperry** [comment]
I was suprised we used wildcards types give the libraries general avoidance of wildcards.
